### PR TITLE
Dlg dot plot

### PR DESCRIPTION
**Pull request by FrancoisJRenaud**
proposal for dlgDotPlot in progress

### CODE REVIEW

**Comment by FrancoisJRenaud** [comment]
> Out of interest, when do we want the sdgPlot to run for this? Do we want it to only run if the sub dialog is opened, or do we want the default options in sdgPlot to run regardless of when it is opened?

I think only when sdg is opened. But what do you mean by "sdgPLots run for this" ?

**Comment by lilyclements** [comment]
Ah don't worry, I checked the sdgPlots and see why you moved it now! I couldn't before because I was on my phone.

**Comment by FrancoisJRenaud** [comment]
> I couldn't before because I was on my phone.
haha, becoming quite geeky over there ;) I hope the weapon is doing ok though ?

File: instat/dlgDotPlot.vb [hunk 1]
```diff
@@ -146,6 +145,8 @@ Public Class dlgDotPlot
     End Sub
 
     Private Sub cmdOptions_Click(sender As Object, e As EventArgs) Handles cmdOptions.Click
+        sdgPlots.SetRSyntax(ucrBase.clsRsyntax)
+        sdgPlots.SetGgplotFunction(clsRggplotFunction)
         sdgPlots.SetDataFrame(strNewDataFrame:=ucrDotPlotSelector.ucrAvailableDataFrames.cboAvailableDataFrames.Text)
```
Comment (maxwellfundi) [inline instat/dlgDotPlot.vb]: Why was this moved here @FrancoisJRenaud . I thought it was agreed we initialize the setup for the subdialogs on the Initialize dialog of the main dialog

Comment (FrancoisJRenaud) [inline instat/dlgDotPlot.vb]: well I don't know who agreed to do this, but creates quite a few bugs. Remember that sdgPlots is shared by many dlg's. If you only set it up in the initialise dialog, sdgPlots will be setup for the last dlg that was initialised, and not for the dlg that has called it last... also see my explanations in #2267 and #2274

Comment (FrancoisJRenaud) [inline instat/dlgDotPlot.vb]: look at how sdgLayerOptions work ... setup is always called right before opening it

**Comment by FrancoisJRenaud** [comment]
This is ready for peer review and testing @maxwellfundi @lilyclements @Lunalo @stevekogo 
I m not super happy with the behaviour of dotplot in R itself. You might find the output of your graphs not as nice as you would like. Don't hesitate to play around with parameters to make it look ok...

**Comment by FrancoisJRenaud** [comment]
Heya ! Just a reminder, could someone review this ? @africanmathsinitiative/instat

File: instat/dlgDotPlot.vb [hunk 2]
```diff
@@ -39,17 +46,17 @@ Public Class dlgDotPlot
...
-        ucrSecondFactorReceiver.Selector = ucrDotPlotSelector
-        ucrSecondFactorReceiver.SetIncludedDataTypes({"factor"})
         ucrBase.clsRsyntax.iCallType = 0
 
-        sdgPlots.SetRSyntax(ucrBase.clsRsyntax)
-        sdgPlots.SetGgplotFunction(clsRggplotFunction)
-        ucrVariablesAsFactorDotPlot.SetFactorReceiver(ucrFactorReceiver)
+        'The ucrVariablesAsFactorDotPlot could be called the ucrBinAxis in spirit.
+        'On the next line, the "Factor" receiver is used as the aesthetics to factor by when different variables are grouped into a single column in the multiple variables method. 
+        ucrVariablesAsFactorDotPlot.SetFactorReceiver(ucrOtherAxisReceiver) 'Could choose the ucrFactorReceiver for this purpose...
```
Comment (lilyclements) [inline instat/dlgDotPlot.vb]: What does .SetFactorReceiver do? Does it mean that it will only accept factors? Or does it mean it will treat any variable inputted as a factor or does it mean something else?

File: instat/dlgDotPlot.vb [hunk 3]
```diff
@@ -66,76 +73,129 @@ Public Class dlgDotPlot
...
-        TestOkEnabled()
+        rdoXBinAxis.Checked = True 'If it was already True, no need to change anything, else it will raise event rdoBinAxis_CheckChanged
+        SetFactorAxisAes()
+        SetBinAxisAes() 'TestOk is done within this sub.
     End Sub
 
-    Private Sub SetXParameter()
-        If ucrFactorReceiver.IsEmpty() = False Then
-            clsRaesFunction.AddParameter("x", ucrFactorReceiver.GetVariableNames(False))
+    Private Sub rdoBinAxis_CheckedChanged(sender As Object, e As EventArgs) Handles rdoXBinAxis.CheckedChanged
```
Comment (lilyclements) [inline instat/dlgDotPlot.vb]: Since this is for when the binAxis' are changed, presumably we should have that it handles when rdoYBinAxis.CheckedChanged as well?

File: instat/dlgDotPlot.vb (hunk 2 above)
Comment (FrancoisJRenaud) [inline instat/dlgDotPlot.vb]: So I just changed the comment about this. As you can see, SetFactorReceiver is a sub from ucrVariablesAsFactorDotplot. Did you check the definition ? Maybe go to definition, check the comment that describes the field it is setting within ucrVariablesAsFactorDotplot and tell me if you understood. If not maybe I can try to make that comment clearer.

File: instat/dlgDotPlot.vb (hunk 3 above)
Comment (FrancoisJRenaud) [inline instat/dlgDotPlot.vb]: I have asked the guys here, and they confirmed that in the case you have only two radiobuttons you only need to handle one of the checkChanged as if one is checked, the other one will be unchecked.

Comment (volloholic) [inline instat/dlgDotPlot.vb]: This is technically correct however radio buttons can also be the source of really annoying bugs. I suggest following lily's advice and consistently handling all the associated radio buttons together would be a good convention for us to follow.  That way if we get bugs they'll all be of the same type and easier to identify and fix throughout the code. 
One obvious way that Radio buttons can lead to bugs is when they're not in a group and you add more radio buttons. We spent half a day debugging a dialog with erratic behaviour caused by hidden ungrouped radio buttons a while back.

**Comment by Lunalo** [comment]
I think we want the counts of Yields within the Vallages. 

I have tested both options and i think working with coord_flip gives a more clear output. 

ggplot(surveyData, aes(x=Village.,y=Yield)) + geom_dotplot(binaxis = "y")+coord_flip()

image

I also tetsed with bin axis and it was not counting the yield by villages.
image

**Comment by FrancoisJRenaud** [comment]
yeah exactly that's my concern. Now the only disadvantage of this is that ideally, the default setting should be "binaxis x". If we work in binaxis y all the time, then that would be achieved by setting coordinate flip as default. Which is not very nice. Any thoughts ? I m happy to have it like said here, or just have default coordinate flip "no" and don't bother about the fact that we then do things along y by default

File: instat/dlgDotPlot.vb (hunk 3 above)
Comment (FrancoisJRenaud) [inline instat/dlgDotPlot.vb]: Oh, should I add a group box as well then ?

**Comment by lilyclements** [comment]
I see. I have a few points I think

I do not think we should have either binaxis or co-ord on the main dialog. I feel the main dialog is for the user to select standard things that will give a default graph.

Now I have explored a little more, I agree that this is the nicest **if** I put my factor in my "Other Axis"
ggplot(survey, aes(y=Yield, x="")) + geom_dotplot(binaxis = "y") + coord_flip()

However, if I put my factor in the "Factor" selector, then having binaxis = "y" and coord_flip()
looks very strange
(I'm using Iris as a dataset and it looks like this)
image

Additionally, if I do two numerical variables and use co-ord_flip and bin = y then both graphs look equally sensible.

**Comment by FrancoisJRenaud** [comment]
I don't think we should not have any choice between axis on the main dialogue, otherwise which defauts do you choose ? Only binaxis x ? But then you get the horrible graphs we are trying to avoid with coord-flip anyway. 
Now about your silly graph above, can you send me the command ? You seem to have y= "" which is not sensible on a dialogue where binaxis is set to y ... the binaxis mapping will be mandatory

**Comment by lilyclements** [comment]
I agree, I don't think we should have choice between axes or co-ord flip on the main dialog.

I did have y= "", the code is
ggplot(data=iris, mapping=aes(y="", x=Petal.Width, fill=Species)) + geom_dotplot(binaxis = "y") + coord_flip()

If I have y = "" and have binaxis = "y" and coord_flip I still feel it looks less sensible than the default
ggplot(data=iris, mapping=aes(y=Sepal.Width, x=Sepal.Length, fill=Species)) + geom_dotplot(binaxis = "y") + coord_flip()

image

My point is that "binaxis = "y"" and "co-ord_flip" are only suitable when you have a factor in your "Other Axis" option so I do not believe we need them to be set as defaults as binaxis = "x" is fine as a default the majority of the time

**Comment by FrancoisJRenaud** [comment]
lets skype I am confused

**Comment by FrancoisJRenaud** [comment]
So @lilyclements, we discussed with @dannyparsons and will just merge things as they are now as it is solid and sensible, although the concerns we have discussed here are still concerns (not final version). I will potentially write an issue summarizing the different possibilities and comments that have been raised. In any case things are still commented inline so it will not be forgotten. People will be able to test see how they feel about the dlg as it is now.
